examples/Attic & Button/app/controllers/demo_controller.rb

```diff
@@ -0,0 +1,266 @@
+            end
+    @discount_percentage = params[:discount_percentage].to_i
+  end
+
+  def payment
+    variation_key, succeeded = @optimizely_service.activate_service!(
+      @current_user,
+      OPTIMIZELY_CONFIG['checkout_flow_experiment']
+    )
+    if succeeded
+      if variation_key
+        @total_price = params[:total_price].to_i
+        session[:checkout_variation_key] = variation_key
+      else
+        flash[:error] = "Failed to create variation using Experiment key: #{OPTIMIZELY_CONFIG['checkout_flow_experiment']}!"
+        redirect_to cart_path(user_id: @current_user['user_id'])
+      end
+    else
+      flash[:error] = @optimizely_service.errors
+      redirect_to cart_path(user_id: @current_user['user_id'])
+    end
+  rescue StandardError => error
+    flash[:error] = "Failed to load datafile using Project ID: #{OPTIMIZELY_CONFIG['project_id']} (#{error})!"
+    redirect_to cart_path(user_id: @current_user['user_id'])
+  end
+
+  def checkout_payment
+    if @optimizely_service.track_service!(
+      OPTIMIZELY_CONFIG['checkout_event_key'],
+      @current_user,
+      'revenue' => params[:total_price].to_i
+    )
+      session[:current_user]['cart'] = []
+      session.delete(:checkout_variation_key)
+      flash[:success] = 'Thank you for shoping'
+      redirect_to shop_path(user_id: @current_user['user_id'])
+    else
+      flash[:error] = @optimizely_service.errors
+      redirect_to payment_path(user_id: @current_user['user_id'])
+    end
+  end
+
+  def log_messages
+    # Returns all log messages
+    @logs = LogMessage.all_logs(@current_user['user_id'])
+  end
+
+  def delete_messages
+    LogMessage.delete_all_logs(@current_user['user_id'])
+    redirect_to messages_path(user_id: @current_user['user_id'])
+    flash[:success] = 'log messages deleted successfully.'
+  end
+
+  def delete_cart
+    session[:current_user]['cart'] = []
+    redirect_to cart_path
+    flash[:success] = 'Cart items successfully removed.'
+  end
+
+  def logout
+    delete_session!
+    redirect_to guest_shop_path
+  end
+
+  def update_cart
+    return render json: {success: false, message: 'Product not found!'} unless @product
+    return render json: {success: false, message: 'Invalid Quantity! valid up to 100'} unless params[:quantity].to_i.between?(0, 100)
+    session[:current_user]['cart'].delete(@product[:id])
+    1.upto(params[:quantity].to_i) { session[:current_user]['cart'] << @product[:id] }
+    render json: {success: true}
+  end
+
+  private
+
+  def initialize_optimizely_client!
+    @optimizely_service = OptimizelyService.new(DATAFILE)
+    @optimizely_service.instantiate! unless OptimizelyService.optimizely_client_present?
+    LogMessage.assign_user!(@current_user['user_id']) if @current_user
+  end
+
+  def delete_session!
+    session.delete(:current_user)
+    session.delete(:variation_key)
+    session.delete(:checkout_variation_key)
```

Comment:
Please also delete the associated log messages

Reply:
Done